random/between.ts

```diff
@@ -0,0 +1,43 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+// This module is browser compatible.
+import type { Prng, RandomOptions } from "./_types.ts";
+export type { Prng, RandomOptions };
+
+/**
+ * Generates a random number between the provided minimum and maximum values.
+ *
+ * The number is in the range `[min, max)`, i.e. `min` is included but `max` is excluded.
+ *
+ * @experimental **UNSTABLE**: New API, yet to be vetted.
+ *
+ * @param min The minimum value (inclusive)
+ * @param max The maximum value (exclusive)
+ * @param options The options for the random number generator
+ * @returns A random number between the provided minimum and maximum values
+ *
+ * @example Usage
+ * ```ts no-assert
+ * import { randomBetween } from "@std/random";
+ *
+ * randomBetween(1, 10); // 6.688009464410508
+ * randomBetween(1, 10); // 3.6267118101712006
+ * randomBetween(1, 10); // 7.853320239013774
+ * ```
+ */
+export function randomBetween(
+  min: number,
+  max: number,
+  options?: RandomOptions,
+): number {
+  if (!Number.isFinite(min)) {
+    throw new RangeError(`min cannot be ${min}`);
+  }
+  if (!Number.isFinite(max)) {
+    throw new RangeError(`max cannot be ${max}`);
+  }
+  if (max < min) {
+    throw new RangeError("max must be greater than or equal to min");
+  }
+
+  return (options?.prng ?? Math.random)() * (max - min) + min;
```

Comment:
This algorithm has a few problems.

- `max - min` can overflow (such as when `min` is `-Number.MAX_VALUE` and `max` is `Number.MAX_VALUE`), making the result either `Infinity` or `NaN` depending on whether the RNG returns `0`.
- Rounding errors can cause the result to sometimes equal `max` (such as when `min` is `1`, `max` is `2` and the RNG returns `0.9999999999999999`) and the distribution to be non-uniform.

A comparably simple alternative (`min * (1 - x) + max * x`) never overflows, but is still prone to rounding errors. [The algorithm used in PHP](https://github.com/php/php-src/blob/407bc092db1729e30eec73147dd0c2a37579ab09/ext/random/gammasection.c#L30-L100) solves both problems but is much more complicated.

Should we change the algorithm or keep the current one?

Reply:
I'm inclined to say this should be fixed even for v0.1 if possible (except for the non-uniform thing, that seems less of a show-stopper as long as the non-uniformity won't cause noticeable bias). But I wonder if there isn't some other way to cheat and get the same effect, e.g. subtracting some multiple of `Number.EPSILON`?

```ts
const x = (options?.prng ?? Math.random)();
return (min * (1 - x) + max * x) - ((max - min) * Number.EPSILON);
```

Reply:
Maybe just map any out-of-bound values to `min`?

```ts
const x = (options?.prng ?? Math.random)();
const y = min * (1 - x) + max * x;
return y >= min && y < max ? y : min;
```